### PR TITLE
Fixed parameter scheduler bug with `CosineAnnealingWarmRestarts`

### LINKED ISSUE #2910: WarmRestarts seems not working with create_lr_scheduler_with_warmup function
## 🐛 Bug description

image

## Environment

- PyTorch Version (e.g., 1.4): 2.0.0+cu118
- Ignite Version (e.g., 0.3.0): 0.4.11
- OS (e.g., Linux): Linux 8082c16db087 `5.10.147+#1` SMP Sat Dec 10 16:00:40 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
- How you installed Ignite (`conda`, `pip`, source): `pip`
- Python version: 3.9.16
- Any other relevant information:

### PR DESCRIPTION
**Pull request by AlexanderChaptykov**
Fixes #2910

Description:

Check list:

- [ ] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

Plotting learning rates:

```python
import matplotlib.pyplot as plt
import numpy as np
import torch
from torch.optim.lr_scheduler import CosineAnnealingWarmRestarts

from ignite.handlers import create_lr_scheduler_with_warmup


def plot(warmup_end_value):
    lr = 0.2
    warm_steps = 5
    steps = 10
    warm_start = 0.023

    def get_optim():
        t1 = torch.zeros([1], requires_grad=True)
        return torch.optim.SGD([t1], lr=lr)

    def get_cos_shed():
        return CosineAnnealingWarmRestarts(optimizer, T_0=12, T_mult=3, verbose=False)

    optimizer = get_optim()
    scheduler = get_cos_shed()
    cosine_lrs = []
    for i in range(steps):
        cosine_lrs.append(optimizer.param_groups[0]["lr"])
        scheduler.step()

    optimizer = get_optim()
    scheduler = create_lr_scheduler_with_warmup(
        get_cos_shed(), warmup_start_value=warm_start, warmup_end_value=warmup_end_value, warmup_duration=warm_steps
    )

    warm_lrs = []
    real_warm_steps = warm_steps if warmup_end_value is not None else (warm_steps - 1)
    for epoch in range(real_warm_steps + steps):
        scheduler(None)
        warm_lrs.append(optimizer.param_groups[0]["lr"])

    if warmup_end_value is not None:
        plt.title("warmup_end_value != lr")
        plt.scatter(range(len(warm_lrs[:real_warm_steps])), warm_lrs[:real_warm_steps])
        plt.scatter(range(warm_steps, len(warm_lrs[real_warm_steps:]) + warm_steps), warm_lrs[real_warm_steps:])
        plt.show()
    else:
        plt.title("warmup_end_value == lr or warmup_end_value is None")
        plt.scatter(range(len(warm_lrs[:warm_steps])), warm_lrs[:warm_steps])
        plt.scatter(range(warm_steps, len(warm_lrs[warm_steps:]) + warm_steps), warm_lrs[warm_steps:])
        plt.show()


plot(None)
plot(.26)
```

image
image

### CODE REVIEW

File: ignite/handlers/param_scheduler.py [hunk 1]
```diff
@@ -852,6 +852,44 @@ def __init__(
...
+                            else:
+                                n = int(math.log((epoch / self.T_0 * (self.T_mult - 1) + 1), self.T_mult))
+                                self.T_cur = epoch - self.T_0 * (self.T_mult**n - 1) / (self.T_mult - 1)
+                                self.T_i = self.T_0 * self.T_mult ** (n)
+                        else:
+                            self.T_i = self.T_0
+                            self.T_cur = epoch
+                    self.last_epoch = math.floor(epoch)
+
+                    return super(_CosineAnnealingWarmRestarts, self).get_lr()
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:883]: Let's define that outside the class, at the same level as other classes.
Let's make it as a wrapping class: `lr_scheduler = _CosineAnnealingWarmRestarts(lr_scheduler)`

Comment (AlexanderChaptykov) [reply ignite/handlers/param_scheduler.py:883]: maybe add `from`  like this?
`lr_scheduler = _CosineAnnealingWarmRestarts.from_scheduler(lr_scheduler)`?

Comment (AlexanderChaptykov) [reply ignite/handlers/param_scheduler.py:883]: why to not declare _CosineAnnealingWarmRestarts inside __init__? 
will we use these two classes (_CosineAnnealingWarmRestarts, LRScheduler) separately from each other?

Comment (vfdev-5) [reply ignite/handlers/param_scheduler.py:883]: > why to not declare _CosineAnnealingWarmRestarts inside __init__?
> will we use these two classes (_CosineAnnealingWarmRestarts, LRScheduler) separately from each other?

No, but code becomes rather hard to read and bulky. Prepending `_` into the class means that it is private, so even defined in the global scope it does not mean that we publicly expose it.

File: ignite/handlers/param_scheduler.py [hunk 2]
```diff
@@ -792,6 +792,69 @@ def simulate_values(  # type: ignore[override]
...
+            else:
+                self._lr_scheduler.T_i = self._lr_scheduler.T_0
+                self._lr_scheduler.T_cur = epoch
+
+        self.last_epoch = math.floor(epoch)
+
+        if not self._get_lr_called_within_step:
+            warnings.warn(
+                "To get the last learning rate computed by the scheduler, " "please use `get_last_lr()`.", UserWarning
+            )
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:847]: I think we can remove this code

File: ignite/handlers/param_scheduler.py [hunk 3]
```diff
@@ -842,7 +905,7 @@ def print_lr():
 
     def __init__(
         self,
-        lr_scheduler: PyTorchLRScheduler,
+        lr_scheduler: Union[PyTorchLRScheduler, _CosineAnnealingWarmRestarts],
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:908]: This should be without `_CosineAnnealingWarmRestarts`
```suggestion
        lr_scheduler: PyTorchLRScheduler,
```

You can annotate `self.lr_scheduler: Union[PyTorchLRScheduler, _CosineAnnealingWarmRestarts] = lr_scheduler` and then handle `isinstance(lr_scheduler, torch.optim.lr_scheduler.CosineAnnealingWarmRestarts)` case.

File: ignite/handlers/param_scheduler.py [hunk 4]
```diff
@@ -792,6 +792,69 @@ def simulate_values(  # type: ignore[override]
             return output
 
 
+class _CosineAnnealingWarmRestarts:
+    def __init__(self, lr_scheduler: torch.optim.lr_scheduler.CosineAnnealingWarmRestarts):
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:796]: Let's directly import `CosineAnnealingWarmRestarts` in the begining of the file to avoid long names like `torch.optim.lr_scheduler.CosineAnnealingWarmRestarts`

File: tests/ignite/handlers/test_param_scheduler.py [hunk 5]
```diff
@@ -1362,3 +1362,33 @@ def test_reduce_lr_on_plateau_scheduler_asserts():
...
+        t1 = torch.zeros([1], requires_grad=True)
+        return torch.optim.SGD([t1], lr=lr)
+
+    def get_cos_shed():
+        return CosineAnnealingWarmRestarts(optimizer, T_0=12, T_mult=3, verbose=False)
+
+    optimizer = get_optim()
+    scheduler = get_cos_shed()
+    cosine_lrs = []
+    for i in range(10):
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_param_scheduler.py:1379]: Let's make more steps, e.g. 100 or 500

File: tests/ignite/handlers/test_param_scheduler.py [hunk 6]
```diff
@@ -1362,3 +1362,33 @@ def test_reduce_lr_on_plateau_scheduler_asserts():
...
+    scheduler = create_lr_scheduler_with_warmup(
+        get_cos_shed(), warmup_start_value=0.023, warmup_end_value=0.23, warmup_duration=10
+    )
+
+    warm_lrs = []
+    for epoch in range(20):
+        scheduler(None)
+        warm_lrs.append(optimizer.param_groups[0]["lr"])
+
+    assert (np.linspace(0.023, 0.23, 10).round(3) == np.array(warm_lrs[:10]).round(3)).all()
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_param_scheduler.py:1393]: Let's introduce some variables for 0.023 and 0.23 values and pytest parametrize the tests with `warmup_end_value` as `[0.23, None]` possible values.

File: ignite/handlers/param_scheduler.py [hunk 7]
```diff
@@ -792,6 +792,69 @@ def simulate_values(  # type: ignore[override]
...
+    def last_epoch(self) -> int:
+        return self._lr_scheduler.last_epoch
+
+    @property
+    def optimizer(self) -> torch.optim.Optimizer:
+        return self._lr_scheduler.optimizer
+
+    @last_epoch.setter
+    def last_epoch(self, value: int):
+        self._lr_scheduler.last_epoch = value
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:809]: Right now, mypy checks are failing here, see https://stackoverflow.com/questions/58121030/how-can-i-use-property-setters-and-make-mypy-happy how to make it happy.
You can test mypy locally with https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md#run-mypy-checks

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the updates @AlexanderChaptykov 
I left few suggestions on how to improve the PR

File: tests/ignite/handlers/test_param_scheduler.py (hunk 6 above)
Comment (AlexanderChaptykov) [reply tests/ignite/handlers/test_param_scheduler.py:1393]: @vfdev-5 could you please provide more info, maybe with examples? How I should parametrize test?

Comment (vfdev-5) [reply tests/ignite/handlers/test_param_scheduler.py:1393]: Here is an example: https://github.com/pytorch/ignite/blob/4a72e07546120e8ded1623d3e15c743ecbe60730/tests/ignite/handlers/test_param_scheduler.py#L380

File: ignite/handlers/param_scheduler.py (hunk 7 above)
Comment (AlexanderChaptykov) [reply ignite/handlers/param_scheduler.py:809]: @vfdev-5 yes, thanks for helping me with this, but how to solve this?
```
ignite/handlers/param_scheduler.py:820: error: Unsupported operand types for * ("int" and "None")  [operator]
                    self._lr_scheduler.T_i = self._lr_scheduler.T_i * self._lr_scheduler.T_mult
```
I can't annotate like this: `self._lr_scheduler.T_mult: int`

File: tests/ignite/handlers/test_param_scheduler.py (hunk 6 above)
Comment (AlexanderChaptykov) [reply tests/ignite/handlers/test_param_scheduler.py:1393]: thanks

File: ignite/handlers/param_scheduler.py [hunk 8]
```diff
@@ -792,6 +792,69 @@ def simulate_values(  # type: ignore[override]
...
+
+    def get_lr(self, epoch: Union[int, None] = None) -> Union[List[float], float]:
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:811]:
```suggestion
    def get_lr(self, epoch: Optional[int] = None) -> List[float]:
```

File: ignite/handlers/param_scheduler.py (hunk 7 above)
Comment (AlexanderChaptykov) [reply ignite/handlers/param_scheduler.py]: update

File: ignite/handlers/param_scheduler.py [hunk 9]
```diff
@@ -852,8 +910,12 @@ def __init__(
                 f"torch.optim.lr_scheduler.{PyTorchLRScheduler.__name__}, "
                 f"but given {type(lr_scheduler)}"
             )
+        self.lr_scheduler: Union[PyTorchLRScheduler, _CosineAnnealingWarmRestarts]
+        if isinstance(lr_scheduler, torch.optim.lr_scheduler.CosineAnnealingWarmRestarts):
+            self.lr_scheduler = _CosineAnnealingWarmRestarts(lr_scheduler)
+        else:
+            self.lr_scheduler = lr_scheduler
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:917]:
```suggestion
        self.lr_scheduler: Union[PyTorchLRScheduler, _CosineAnnealingWarmRestarts] = lr_scheduler
        if isinstance(lr_scheduler, CosineAnnealingWarmRestarts):
            self.lr_scheduler = _CosineAnnealingWarmRestarts(lr_scheduler)
```

Comment (AlexanderChaptykov) [reply ignite/handlers/param_scheduler.py:917]: that is much better, respect

File: tests/ignite/handlers/test_param_scheduler.py [hunk 10]
```diff
@@ -1362,3 +1362,44 @@ def test_reduce_lr_on_plateau_scheduler_asserts():
...
+        warm_lrs.append(optimizer.param_groups[0]["lr"])
+
+    if warmup_end_value is not None:
+        assert (
+            np.linspace(warm_start, warmup_end_value, warm_steps).round(3) == np.array(warm_lrs[:warm_steps]).round(3)
+        ).all()
+        assert warm_lrs[warm_steps:] == cosine_lrs
+    else:
+        assert (np.linspace(warm_start, lr, warm_steps).round(3) == np.array(warm_lrs[:warm_steps]).round(3)).all()
+        assert warm_lrs[warm_steps - 1 : -1] == cosine_lrs
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_param_scheduler.py:1405]:
```suggestion
        assert warm_lrs[warm_steps - 1:] == cosine_lrs
```

File: tests/ignite/handlers/test_param_scheduler.py [hunk 11]
```diff
@@ -1362,3 +1362,44 @@ def test_reduce_lr_on_plateau_scheduler_asserts():
...
+
+    warm_lrs = []
+    for epoch in range(warm_steps + steps):
+        scheduler(None)
+        warm_lrs.append(optimizer.param_groups[0]["lr"])
+
+    if warmup_end_value is not None:
+        assert (
+            np.linspace(warm_start, warmup_end_value, warm_steps).round(3) == np.array(warm_lrs[:warm_steps]).round(3)
+        ).all()
```
[the remark this answers is not in the document]
Comment (AlexanderChaptykov) [reply tests/ignite/handlers/test_param_scheduler.py:1401]: without `round`:
array([ True,  True,  True,  True,  True, False,  True,  True,  True,
        True,  True, False,  True,  True,  True,  True,  True,  True,
        True, False,  True,  True, False,  True,  True,  True,  True,
       False,  True,  True, False,  True,  True, False,  True, False,
        True,  True, False,  True, False,  True,  True, False, False,
       False,  True,  True,  True,  True])

[the remark this answers is not in the document]
Comment (vfdev-5) [reply tests/ignite/handlers/test_param_scheduler.py:1401]: please use `np.testing.assert_allclose(np.linspace(warm_start, warmup_end_value, warm_steps), warm_lrs[:warm_steps])`

File: tests/ignite/handlers/test_param_scheduler.py (hunk 10 above)
Comment (AlexanderChaptykov) [reply tests/ignite/handlers/test_param_scheduler.py:1405]: we need this, beacuse of shifting lrs if warmup_end_value == None

**Comment by vfdev-5** [comment]
Let's make the test as following:
```python
@pytest.mark.parametrize("warmup_end_value", [0.23, None])
@pytest.mark.parametrize("T_0", [1, 12])
@pytest.mark.parametrize("T_mult", [1, 3])
def test_create_lr_scheduler_with_warmup_cosine(warmup_end_value, T_0, T_mult):
    lr = 0.2
    steps = 200
    warm_steps = 50
    warm_start = 0.023

    def get_optim():
        t1 = torch.zeros([1], requires_grad=True)
        return torch.optim.SGD([t1], lr=lr)

    def get_cos_shed():
        return CosineAnnealingWarmRestarts(optimizer, T_0=T_0, T_mult=T_mult, verbose=False)

    optimizer = get_optim()
    scheduler = get_cos_shed()
    cosine_lrs = []
    for i in range(steps):
        cosine_lrs.append(optimizer.param_groups[0]["lr"])
        scheduler.step()

    optimizer = get_optim()
    scheduler = create_lr_scheduler_with_warmup(
        get_cos_shed(), warmup_start_value=warm_start, warmup_end_value=warmup_end_value, warmup_duration=warm_steps
    )

    warm_lrs = []
    for epoch in range(warm_steps + steps):
        scheduler(None)
        warm_lrs.append(optimizer.param_groups[0]["lr"])

    if warmup_end_value is not None:
        np.testing.assert_allclose(np.linspace(warm_start, warmup_end_value, warm_steps), warm_lrs[:warm_steps])
        assert warm_lrs[warm_steps:] == cosine_lrs
    else:
        np.testing.assert_allclose(np.linspace(warm_start, lr, warm_steps), warm_lrs[:warm_steps])
        assert warm_lrs[warm_steps - 1:-1] == cosine_lrs
```

File: tests/ignite/handlers/test_param_scheduler.py (hunk 10 above)
Comment (vfdev-5) [reply tests/ignite/handlers/test_param_scheduler.py:1405]: actually, it would be better to register the same number of samples instead of slicing here, otherwise it is unlear

**Comment by vfdev-5** [comment]
```python
import matplotlib.pyplot as plt
import numpy as np
import torch
from torch.optim.lr_scheduler import CosineAnnealingWarmRestarts

from ignite.handlers import create_lr_scheduler_with_warmup


def plot(warmup_end_value):
    lr = 0.2
    warm_steps = 5
    steps = 100
    warm_start = 0.023

    def get_optim():
        t1 = torch.zeros([1], requires_grad=True)
        return torch.optim.SGD([t1], lr=lr)

    def get_cos_shed():
        return CosineAnnealingWarmRestarts(optimizer, T_0=10, T_mult=2, verbose=False)

    optimizer = get_optim()
    scheduler = get_cos_shed()
    cosine_lrs = []
    for i in range(steps):
        cosine_lrs.append(optimizer.param_groups[0]["lr"])
        scheduler.step()

    optimizer = get_optim()
    scheduler = create_lr_scheduler_with_warmup(
        get_cos_shed(), warmup_start_value=warm_start, warmup_end_value=warmup_end_value, warmup_duration=warm_steps
    )

    warm_lrs = []
    for epoch in range(warm_steps + steps):
        scheduler(None)
        warm_lrs.append(optimizer.param_groups[0]["lr"])

    if warmup_end_value is not None:
        plt.figure(figsize=(10, 5))
        plt.subplot(121)
        plt.title("create_lr_scheduler_with_warmup +\nCosineAnnealingWarmRestarts\nwarmup_end_value != lr")
        plt.plot(warm_lrs, "-*")
        plt.subplot(122)
        plt.title("CosineAnnealingWarmRestarts")
        plt.plot(cosine_lrs, "-*")        
        plt.show()
    else:
        plt.figure(figsize=(10, 5))
        plt.subplot(121)
        plt.title("create_lr_scheduler_with_warmup +\nCosineAnnealingWarmRestarts\nwarmup_end_value == lr")
        plt.plot(warm_lrs, "-*")
        plt.subplot(122)
        plt.title("CosineAnnealingWarmRestarts")
        plt.plot(cosine_lrs, "-*")        
        plt.show()


plot(None)
plot(.26)
```

image
image
